android-json-form-wizard/src/main/java/com/vijay/jsonwizard/widgets/GpsFactory.java

```diff
@@ -224,27 +223,44 @@ protected void addViewTags(Button recordButton, WidgetArgs widgetArgs, WidgetMet
         recordButton.setTag(R.id.extraPopup, widgetArgs.isPopup());
     }
 
-    public void attachLayout(Context context, @NonNull JSONObject jsonObject, @NonNull View dataView, @NonNull TextView latitudeTv, @NonNull TextView longitudeTv, @NonNull TextView altitudeTv, @NonNull TextView accuracyTv) {
-        String latitude = "";
-        String longitude = "";
-        String accuracy = "";
-        String altitude = "";
+    public void attachLayout(Context context, @NonNull JSONObject jsonObject, @NonNull View dataView,
+                             @NonNull TextView latitudeTv, @NonNull TextView longitudeTv,
+                             @NonNull TextView altitudeTv, @NonNull TextView accuracyTv) {
+        TextView[] views = new TextView[]{latitudeTv, longitudeTv, altitudeTv, accuracyTv};
+
+        int[] stringIds = new int[]{R.string.latitude, R.string.longitude, R.string.altitude, R.string.accuracy};
+
         if (jsonObject.has(JsonFormConstants.VALUE)) {
             String coordinateData = jsonObject.optString(JsonFormConstants.VALUE);
 
             String[] coordinateElements = coordinateData.split(" ");
-            if (coordinateElements.length > 1) {
-                latitude = coordinateElements[0];
-                longitude = coordinateElements[1];
+
+            //loops to fill textViews with values supplied
+            for (int i = 0; i < coordinateElements.length; i++) {
+                views[i].setText(getText(context, coordinateElements[i], stringIds[i]));
             }
-        }
 
-        latitudeTv.setText(getText(context, latitude, R.string.latitude));
-        longitudeTv.setText(getText(context, longitude, R.string.longitude));
-        altitudeTv.setText(getText(context, altitude, R.string.altitude));
-        accuracyTv.setText(getText(context, accuracy, R.string.accuracy));
+            //loops to fill empty textViews with empty string since the values is not available
+            for (int i = coordinateElements.length; i < stringIds.length; i++) {
+                views[i].setText(getText(context, "", stringIds[i]));
+            }
 
-        dataView.setTag(R.id.raw_value, constructString(latitude, longitude));
+            dataView.setTag(R.id.raw_value, constructString(coordinateElements));
+        } else {
+            //loops to fill all textViews with empty string
+            for (int i = 0; i < stringIds.length; i++) {
+                views[i].setText(getText(context, "", stringIds[i]));
+            }
+            dataView.setTag(R.id.raw_value, "");
```

Comment:
Is there a reason we don't have the text set to empty by default? We end up repeating the same logic here https://github.com/OpenSRP/opensrp-client-native-form/pull/524/files#diff-607b2f74eb9df7c9ec407f98d71dc161ce12967b037b971ddcc4564b8ff5fefeR244 right?

Reply:
you mean on the `xml`?  if yes the text attribute of the `TextView` has been assigned a string resource id ; if no we will still need to call `.setText()` for each of those views

Reply:
Why do you need to call setText twice if it's being set to empty string? Can't it be done once initially (as a default) and let the values be overwritten if data for them exists?

Also, couldn't the xml have the default value already set?

Reply:
`Why do you need to call setText twice if it's being set to empty string?` its not called twice there is an else

Reply:
Isn't the same logic done in both the if and else?

Reply:
Yeah its done in `if` only if the value has fewer coordinate elements than the number of textViews ... its done on else only when the value attribute does not exist

Reply:
am only setting to empty when needed

Reply:
Wouldn't it be cleaner to have the defaults already in place and not have to add an else or to worry about checking if the coordinates were fewer?

Reply:
Maybe but ... just thought calling `.setText` twice for all the views i.e one for setting empty text and the other for actual text, will be unnecessary.

Reply:
The ideal way would be to have the default text in xml, is this not possible?

Either way I think the benefit of having concise and non-repeated code is a win here since the setText operation is not expensive.